### PR TITLE
apt yum: use fluentd.service by default

### PR DESCRIPTION
**Pull request by kenhys**
Basically, fluentd.service should be used by default, but with considering
migration from old td-agent, added alias (Alias=td-agent.service)
in fluentd.service.

To distinguish service name and compatible service name,
it use SERVICE_NAME (fluentd) and COMPAT_SERVICE_NAME (td-agent) respectively.

When upgrading from v4 package, td-agent service will be removed by
default, but instead of it, fluentd.service launches.

This limitation is caused by execution order of old package's
 maintainer scriptlet behavior (%postun).

If you want to use td-agent as well, you must re-enable service via
systemctl explicitly.

deb:
  sudo systemctl unmask td-agent
  sudo systemctl enable fluentd
(td-agent.service is masked, so need to unmask first)

rpm:
  sudo systemctl enable fluentd
(td-agent.service is inactive, so need to enable first)

NOTE: service environment file is moved to /etc/default/fluentd or
/etc/sysconfig/fluentd for consistency.

### CODE REVIEW

**Comment by kenhys** [comment]
Chekcking diff outcome.

```
diff deb
--- fluent5deb.orig     2023-05-11 16:51:26.598363453 +0900
+++ fluent5deb.rev      2023-05-11 16:51:26.878363842 +0900
@@ -7,7 +7,7 @@
 ./lib/
 ./lib/systemd/
 ./lib/systemd/system/
-./lib/systemd/system/td-agent.service
+./lib/systemd/system/fluentd.service
 ./opt/
 ./opt/td-agent/
 ./opt/td-agent/LICENSES/
diff rpm
--- fluent5rpm.orig     2023-05-11 16:51:26.930363913 +0900
+++ fluent5rpm.rev      2023-05-11 16:51:26.978363979 +0900
@@ -9362,7 +9362,7 @@
 /opt/td-agent/share/td-agent.conf.tmpl
 /tmp/td-agent
 /usr/bin/td
-/usr/lib/systemd/system/td-agent.service
+/usr/lib/systemd/system/fluentd.service
 /usr/lib/tmpfiles.d/td-agent.conf
 /usr/sbin/td-agent
 /usr/sbin/td-agent-gem
```

**Comment by ashie** [comment]
> If you use RPM, Alias=td-agent.service is set to
fluentd.service, but it is effective for fresh install only.
Upgraded from old package, you must systemctl enable fluentd
explicitly.(as old package's %postun removed td-agent.service)

Oops, it's not desired. Should we swap entity & alias, or add something other to resolve it?

**Comment by kenhys** [comment]
```
 ./pkgdiff.sh fluent
diff deb
--- fluent5deb.orig     2023-05-12 14:39:17.331699815 +0900
+++ fluent5deb.rev      2023-05-12 14:39:17.627699196 +0900
@@ -7,7 +7,7 @@
 ./lib/
 ./lib/systemd/
 ./lib/systemd/system/
-./lib/systemd/system/td-agent.service
+./lib/systemd/system/fluentd.service
 ./opt/
 ./opt/td-agent/
 ./opt/td-agent/LICENSES/
diff rpm
--- fluent5rpm.orig     2023-05-12 14:39:17.675699095 +0900
+++ fluent5rpm.rev      2023-05-12 14:39:17.723698993 +0900
@@ -1,5 +1,5 @@
 /etc/logrotate.d/td-agent
-/etc/sysconfig/td-agent
+/etc/sysconfig/fluentd
 /etc/td-agent
 /etc/td-agent/plugin
 /etc/td-agent/td-agent.conf
@@ -9362,7 +9362,7 @@
 /opt/td-agent/share/td-agent.conf.tmpl
 /tmp/td-agent
 /usr/bin/td
-/usr/lib/systemd/system/td-agent.service
+/usr/lib/systemd/system/fluentd.service
 /usr/lib/tmpfiles.d/td-agent.conf
 /usr/sbin/td-agent
 /usr/sbin/td-agent-gem
```

**Comment by kenhys** [comment]
TODO:
Use modern macros.
https://docs.fedoraproject.org/en-US/packaging-guidelines/Scriptlets/#_systemd

**Comment by kenhys** [comment]
Fixed to use modern macro.

**Comment by kenhys** [comment]
waiting CI results.

**Review by ashie — APPROVED** [review]
> This limitation is caused by execution order of old package's
> %postun maintainer script behavior. (no workaround)

There is a hacky work around to resolve it.
`%posttrans` of the new package will be run after `%postun` of the old one.
So that running `systemctl enable fluentd` at `%posttrans` will resolve it.

But I think we don't have to go that far to enable it.
We don't need to add such hacky way in this PR.
If it's really needed, we'll do it in another PR.

**Review by ashie — CHANGES_REQUESTED** [review]
Hmm, I can't migrate properly on Ubuntu.
We might need more work.

**Comment by kenhys** [comment]
> We might need more work.

Sure, I'll fix it.

**Comment by kenhys** [comment]
On ubuntu side, dh_installsystemd assumes service_name instead of compat_service_name.

**Comment by ashie** [comment]
Thanks, I've tested it again on Ubuntu.
Now fluentd service is successfully launched after upgrade!
Although td-agent service is still masked, it's acceptable for me.
On the contrary, it has the advantage of letting the user know that the service name has changed.
(Of course we should describe how to upgrade in the document.)

**Comment by ashie** [comment]
me too

File: fluent-package/yum/fluent-package.spec.in [hunk 1]
```diff
@@ -139,29 +138,31 @@ mkdir -p %{buildroot}%{_sysconfdir}/@PACKAGE_DIR@/plugin
...
-if [ $1 -eq 2 ]; then
-  %{control_service condrestart @SERVICE_NAME@}
+    systemctl is-active @COMPAT_SERVICE_NAME@
+    if [ $? -eq 0 ]; then
+	# When upgrading from v4, td-agent.service will be removed
+	# with %postun scriptlet. fluentd service also inactive even though
+	# td-agent.service is running before upgrade process. Try to
+	# keep running fluentd service, explicitly restart it.
+	systemctl restart @SERVICE_NAME@.service
+    fi
```
Comment (ashie) [inline fluent-package/yum/fluent-package.spec.in:165]: Is it also workable for upgrading to upper version of fluent-package?
(I'm not sure, will try from now.)

Comment (kenhys) [reply fluent-package/yum/fluent-package.spec.in:165]: I think so.

* upgrade from v4 => td-agent inactive, fluentd active
* upgrade from 5.0.0 (td-agent inactive) => td-agent inactive, fluentd active
* upgrade from 5.0.0 (td-agent active) => td-agent active, fluentd active

**Comment by kenhys** [comment]
Updated commit message. f531fa5

File: fluent-package/yum/fluent-package.spec.in (hunk 1 above)
Comment (ashie) [reply fluent-package/yum/fluent-package.spec.in:165]: I've created v5.0.1 and test it.
It also works fine. Thanks!

**Review by ashie — APPROVED** [review]
LGTM

Waiting for CI...

**Comment by ashie** [comment]
Thanks a lot!

**Pull request merged**